content/redirect.js

```diff
@@ -2,18 +2,68 @@ const fs = require("fs");
 const path = require("path");
 
 const { resolveFundamental } = require("../libs/fundamental-redirects");
-const { CONTENT_ROOT, VALID_LOCALES } = require("./constants");
+const { decodePath, slugToFolder } = require("../libs/slug-utils");
+const {
+  CONTENT_ROOT,
+  CONTENT_TRANSLATED_ROOT,
+  VALID_LOCALES,
+} = require("./constants");
+
+const FORBIDDEN_URL_SYMBOLS = ["\n", "\t"];
+
+let ARCHIVED_URLS;
+
+function checkURLInvalidSymbols(url) {
+  for (const character of FORBIDDEN_URL_SYMBOLS) {
+    if (url.includes(character)) {
+      throw new Error(`URL contains invalid character '${character}'`);
+    }
+  }
+}
+
+function documentExists(url) {
+  // Let's keep vanity urls to /en-US/
+  if (url === "/en-US/") {
+    return url;
+  }
+  const [bareURL] = url.split("#", 2);
+
+  if (!ARCHIVED_URLS) {
```

Comment:
Nit: instead of switching types, perhaps make it a Set always and only extend it if `if (!ARCHIVED_URLS.size)`.